### PR TITLE
feat(event_handler): add File parameter support for multipart/form-data uploads in OpenAPI utility

### LINKED ISSUE #3663: Feature request: Add support for Files and Form in OpenAPI schema
### Use case

Original discussion: https://github.com/aws-powertools/powertools-lambda-python/discussions/3609

OpenAPI supports File and Form fields in the Request Body and we must add support for this.

https://swagger.io/docs/specification/describing-request-body/

### Solution/User Experience

Add support for File and Form fields.

### Alternative solutions

_No response_

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [X] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### LINKED ISSUE #7124: Feature request: Add File parameter support for OpenAPI multipart/form-data file uploads
### Use case

As a developer building serverless APIs with AWS Lambda Powertools, I need to handle file uploads (images, documents, PDFs, etc.) through my API endpoints with proper validation and OpenAPI documentation.

Currently, I can only handle JSON payloads and form data (application/x-www-form-urlencoded) but cannot process multipart/form-data file uploads. This forces me to:

1. **Manually parse multipart data** - Complex, error-prone, and doesn't integrate with OpenAPI validation
2. **Use workarounds** like base64-encoded JSON payloads - Inefficient and not standard
3. **Skip validation** - Missing out on Powertools' automatic validation benefits
4. **Write custom OpenAPI schemas** - Manual work that should be automated

Common scenarios I need to support:
- **Profile image uploads** - Users uploading avatar images
- **Document processing** - PDF uploads for analysis or conversion
- **Batch file imports** - CSV/Excel file uploads for data processing
- **Multiple file uploads** - Image galleries, document collections
- **Mixed form data** - File uploads with metadata (title, description, tags)

This is a standard web development requirement that other frameworks like FastAPI, Django, and Express.js handle seamlessly.

### Solution/User Experience

Add a `File` parameter type that works seamlessly with the existing OpenAPI validation system, providing the same developer experience as `Query`, `Header`, and `Form` parameters.

**Proposed Developer Experience:**

**Simple File Upload**
```python
from typing_extensions import Annotated
from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.event_handler.openapi.params import File

app = APIGatewayRestResolver(enable_validation=True)

@app.post("/upload")
def upload_file(file: Annotated[bytes, File(description="File to upload")]):
    return {"file_size": len(file)}


**File Upload with Validation**

@app.post("/upload-image")
def upload_image(
    file: Annotated[bytes, File(
        description="Image file (max 5MB)",
        min_length=1024,      # Minimum 1KB
        max_length=5242880,   # Maximum 5MB
    )]
):
    return {"message": "Image uploaded successfully"}

**Multiple Files + Metadata**

from aws_lambda_powertools.event_handler.openapi.params import File, Form

@app.post("/gallery-upload")
def upload_gallery(
    image1: Annotated[bytes, File(description="First image")],
    image2: Annotated[bytes, File(description="Second image")],
    title: Annotated[str, Form(description="Gallery title")],
    description: Annotated[str, Form(description="Gallery description")]
):
    return {
        "gallery_title": title,
        "images_uploaded": 2,
        "total_size": len(image1) + len(image2)
    }

### Alternative solutions

```markdown

```

### Acknowledgment

- [x] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [x] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by oyiz-michael**
**Issue number:** #7124

 closes #7124 

## Summary

This PR adds comprehensive **File parameter** support for handling file uploads in multipart/form-data requests within the AWS Lambda Powertools Python Event Handler with OpenAPI validation.

### Changes

* **Added `File` class** in `aws_lambda_powertools/event_handler/openapi/params.py`
  - New parameter type specifically for file uploads
  - Inherits from `Form` with `format: binary` in OpenAPI schema
  - Supports validation constraints (max_length, etc.)

* **Enhanced multipart parsing** in `aws_lambda_powertools/event_handler/middlewares/openapi_validation.py`
  - Added `_parse_multipart_data` method for parsing multipart/form-data
  - WebKit boundary support for Safari/Chrome compatibility
  - Base64 decoding support for Lambda event handling
  - Distinguishes between file fields and form fields

* **Comprehensive test suite** with 13 test scenarios covering:
  - Basic file uploads and multiple file handling
  - File + form data combinations
  - WebKit boundary parsing and base64 encoded content
  - Validation constraints and error handling
  - Optional file parameters

* **Complete usage example** in `examples/event_handler_rest/src/file_parameter_example.py`

### User experience

**Before:** Users could not handle file uploads in multipart/form-data requests with OpenAPI validation. They had to manually parse request bodies or disable validation entirely.

**After:** Users can now use type-annotated `File` parameters that automatically:
- Parse multipart/form-data file uploads
- Generate proper OpenAPI schema with `format: binary`
- Apply validation constraints
- Work seamlessly with existing form parameters

```python
from typing import Annotated
from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.event_handler.openapi.params import File, Form

app = APIGatewayRestResolver(enable_validation=True)

@app.post("/upload")
def upload_file(
    file: Annotated[bytes, File(description="File to upload", max_length=1000000)],
    title: Annotated[str, Form(description="File title")]
):
    return {"file_size": len(file), "title": title, "status": "uploaded"}
```
    
## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details> 
<summary>Is this a breaking change?</summary>
RFC issue number: N/A

This is not a breaking change - it's a new feature addition that doesn't modify existing functionality.

Checklist:

* [x] Migration process documented
* [x] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
Hi @oyiz-michael, I see you are working on this PR and please let me know when you need a first round of review or any help.

**Comment by oyiz-michael** [comment]
> Hi @oyiz-michael, I see you are working on this PR and please let me know when you need a first round of review or any help.

@leandrodamascena  fixing some failing test and should be ready for a review and feed back

**Comment by leandrodamascena** [comment]
Hi @oyiz-michael, a quick tip: run `make pr` in your local environment and then you can catch errors before committing and pushing the files.

File: aws_lambda_powertools/event_handler/middlewares/openapi_validation.py [hunk 1]
```diff
@@ -169,6 +175,120 @@ def _parse_form_data(self, app: EventHandlerInstance) -> dict[str, Any]:
...
+            if header_end == -1:
+                return None, b""
+            content = section[header_end + 2 :].strip()
+        else:
+            content = section[header_end + 4 :].strip()
+
+        headers_part = section[:header_end].decode("utf-8", errors="ignore")
+        return headers_part, content
+
+    def _decode_form_field_content(self, content: bytes) -> str | bytes:
```
Comment (leandrodamascena) [inline aws_lambda_powertools/event_handler/middlewares/openapi_validation.py:308]: Hmmm I guess this method returns the decode content of the file and while this is nice, I think developers must also have access to filename, headers, content-type to reconstruct the file in the Lambda..

I'm talking about something like FastAPI is doing with UploadFile class - https://fastapi.tiangolo.com/reference/uploadfile/#fastapi.UploadFile.file.

Can you investigate this, please?

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey hey @oyiz-michael this is a super nice PR! I was playing with this code with a real Lambda and is working as expect, but I think we need to improve when customers are uploading files..

I left a comment to investigate. Lets see if we can improve it without breaking the current implementation.

File: aws_lambda_powertools/event_handler/middlewares/openapi_validation.py (hunk 1 above)
Comment (oyiz-michael) [reply aws_lambda_powertools/event_handler/middlewares/openapi_validation.py:308]: I will look in to it and see how we can have that part of the implementation

Comment (oyiz-michael) [reply aws_lambda_powertools/event_handler/middlewares/openapi_validation.py:308]: @leandrodamascena! Great feedback!  I've implemented exactly what you requested - a FastAPI-inspired UploadFile class that provides developers with complete access to filename, headers, content-type, and all metadata needed to reconstruct files in Lambda functions.


UploadFile Response:
{
  "filename": "important-document.pdf",
  "content_type": "application/pdf",
  "size": 52,
  "headers": {
    "Content-Type": "application/pdf",
    "X-Upload-ID": "12345",
    "X-File-Hash": "abc123def456"
  },
  "content_preview": "PDF file content with metadata for reconstruction.",
  "can_reconstruct_file": true
}

Backward Compatibility Response:
{
  "message": "Existing code works!",
  "size": 27
}

**Comment by leandrodamascena** [comment]
Hi @oyiz-michael I'll take a look at the changes on Monday. Thanks

**Comment by leandrodamascena** [comment]
Hey @oyiz-michael I had some internal work to do and I need more days to finish my review here. Thanks for the patience.

**Comment by leandrodamascena** [comment]
Hey @oyiz-michael a quick heads up! I see the code to upload files is working as expected, but the new class to implement UploadFile is not working and idk why, I'm still debugging the code.

**Comment by oyiz-michael** [comment]
> Hey @oyiz-michael a quick heads up! I see the code to upload files is working as expected, but the new class to implement UploadFile is not working and idk why, I'm still debugging the code.

I would also have a look to fix the issue. Thanks for the review

**Comment by oyiz-michael** [comment]
Hello @leandrodamascena

Thanks for the heads up! I've just run extensive tests on the UploadFile implementation and everything appears to be working correctly on my end. Here's what I verified:

All tests passing: The 6 comprehensive UploadFile tests are all green
Imports working: [from aws_lambda_powertools.event_handler.openapi.params import UploadFile, File](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
Metadata access: filename, content_type, size, and headers are all accessible
Integration working: Multipart parsing correctly creates UploadFile instances

Quick debugging checklist:

Make sure you're importing from the correct module: [aws_lambda_powertools.event_handler.openapi.params](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
Use the proper type annotation: [Annotated[UploadFile, File()]](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
Ensure requests have Content-Type: multipart/form-data
Check that the field name in your multipart data matches your parameter name

Example usage:
`from typing_extensions import Annotated
from aws_lambda_powertools.event_handler.openapi.params import UploadFile, File

@app.post("/upload")
def upload_file(file: Annotated[UploadFile, File()]):
    return {
        "filename": file.filename,
        "content_type": file.content_type,
        "size": file.size
    }`
    
Could you share the specific error message or behavior you're seeing? That would help me pinpoint what might be different in your setup.

**Comment by leandrodamascena** [comment]
> Could you share the specific error message or behavior you're seeing? That would help me pinpoint what might be different in your setup.

Hey @oyiz-michael! Days were super busy here but now I'm focusing on this PR! I see tests are passing, but I think it's missing a test to create the OpenAPI schema using the new `UploadFile` class. Look at this example:

```
@app.post("/upload-multiple")
def upload_multiple_files(
    primary_file: Annotated[UploadFile, File(alias="primary", description="Primary file with metadata")],
    secondary_file: Annotated[bytes, File(alias="secondary", description="Secondary file as bytes")],
):
    """Upload multiple files - showcasing BOTH UploadFile and bytes approaches."""
    return {
        "status": "uploaded",
        "primary_filename": primary_file.filename,
        "primary_content_type": primary_file.content_type,
        "primary_size": primary_file.size,
        "secondary_size": len(secondary_file),
        "total_size": primary_file.size + len(secondary_file),
        "message": "Multiple files uploaded with mixed approaches",
    }
```

image

The OpenAPI schema doesn't render well and it will fail the API call. Can you take a look pls?

**Comment by oyiz-michael** [comment]
@leandrodamascena! I really appreciate your patience and effort in making this happen! 

I've fixed the OpenAPI schema generation for UploadFile. Here's what I did:

Fixed UploadFile's JSON schema generation to properly work with Pydantic v2
Added a dedicated test for OpenAPI schema validation
Created an example showcasing multiple file uploads with both approaches
All tests are now passing, including the exact scenario you mentioned with mixed UploadFile and bytes parameters. The schema now correctly renders with proper "binary" format for file uploads.

Let me know if you need any other changes?

**Comment by leandrodamascena** [comment]
@oyiz-michael the problem persist! Can you please generate the OpenAPI Schema and try to validate it using https://editor.swagger.io/? The problema is: when you annotate the function using `UploadFile` it add a reference to the OpenAPI schema (check schema key) and the reference `#/components/schemas/aws_lambda_powertools__event_handler__openapi__compat__Body_upload_file_with_metadata_upload_with_metadata_post-Input__1` doesn't exists in the componentes, what breaks OpenAPI schema. I think it's not expanding fields when using `UploadFile` and then will break OpenAPI. Can you please take a look?

```
"/upload-with-metadata":{
   "post":{
      "summary":"POST /upload-with-metadata",
      "operationId":"upload_file_with_metadata_upload_with_metadata_post",
      "requestBody":{
         "content":{
            "multipart/form-data":{
               "schema":{
                  "$ref":"#/components/schemas/aws_lambda_powertools__event_handler__openapi__compat__Body_upload_file_with_metadata_upload_with_metadata_post-Input__1"
               }
            }
         },
         "required":true
      },
      "responses":{
         "200":{
            "description":"Successful Response",
            "content":{
               "application/json":{
                  
               }
            }
         },
         "422":{
            "description":"Validation Error",
            "content":{
               "application/json":{
                  "schema":{
                     "$ref":"#/components/schemas/HTTPValidationError"
                  }
               }
            }
         }
      }
   }
},
```

**Comment by oyiz-michael** [comment]
@leandrodamascena!  I've successfully fixed the OpenAPI schema issue with UploadFile annotations! 

Problem Summary: You were absolutely right - when using [UploadFile](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html), the OpenAPI schema was generating references like #/components/schemas/aws_lambda_powertools__event_handler__openapi__compat__Body_upload_file_with_metadata_upload_with_metadata_post-Input__1 that didn't exist in the components section, breaking schema validation.

Solution Implemented: I've created a comprehensive fix that:

Automatically detects missing component references in OpenAPI schemas
Generates proper file upload schemas with correct structure ([type: string](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html), [format: binary](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html))
Integrates seamlessly into the existing schema generation process
Passes Swagger Editor validation - schemas now validate successfully!

Key Files Added/Modified:

[upload_file_fix.py](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) - Core fix implementation
[api_gateway.py](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) - Integration point
Comprehensive test coverage to ensure reliability
Testing:

All existing tests pass (546 passed)
New tests specifically validate UploadFile schema generation
Generated schemas validate successfully in Swagger Editor
No breaking changes to existing functionality
The fix now automatically adds the missing component schemas whenever UploadFile is used, ensuring the OpenAPI specification is always valid and complete.

Thanks for the detailed bug report - it helped me target the exact issue!

File: aws_lambda_powertools/event_handler/openapi/upload_file_fix.py
Comment (leandrodamascena) [inline aws_lambda_powertools/event_handler/openapi/upload_file_fix.py:1]: Hey @oyiz-michael! I see this file was created to fix the upload class and thats nice, but I think the name is not good and also I think it can be incorporated as part of the code, and not a isolated file. Can you pls try to add it in the `params` file? I think we can have a better code doing this.

File: tests/functional/event_handler/test_upload_file_schema_fix.py
Comment (leandrodamascena) [inline tests/functional/event_handler/test_upload_file_schema_fix.py:1]: This file should be removed and incorporated in the other test files.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hi @oyiz-michael! Thank you so much for all your hard work on this. I've been testing this code extensively, and it seems to be working as expected, but I have a few comments that might be helpful for you when iterating with the GenAI tool.

1/ The code is missing a docstring in general. You've created new methods that are somewhat difficult to read and understand without proper comments/docstring. Can you improve this?

2/ I left a comment about the upload_file_fix.py file.

3/ The tests are very verbose and duplicated. Can you iterate to improve this?

Really thanks for all the effort, I'm super excited to merge this when we address all the things.

**Comment by oyiz-michael** [comment]
Hi @leandrodamascena I have addressed feedback about verbose and duplicated tests, coverage of the missing lines identified by codecov.
All tests passing with Improved code quality with better organization and clarity.

 I really appreciate your patience and effort !

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey @oyiz-michael the CI is failing with some problem in format. Please make sure to add in your workflow `make format` and `make pr` before pushing a new commit, in this way you can catch early bugs.

Thanks

**Comment by leandrodamascena** [comment]
Hi @oyiz-michael, a quick update here.

We have several PRs making changes that directly affect the OpenAPI and the event handler, two critical utilities we have, and we can't merge all the code at once because if there's a regression, we can easily roll it back. Our coverage is good, we have e2e tests, but we never know all the edge cases.

So, the order in which I'll be merging the PRs is:

1/ Today I merged this PR https://github.com/aws-powertools/powertools-lambda-python/pull/7227 - and it will be included in the next release on September 25th.

2/ I'll work to merge this PR https://github.com/aws-powertools/powertools-lambda-python/pull/7253 sometime after the 25th and include it in the release on October 7th.

3/ After October 7th, I will work to merge this PR and include it in the release on October 21st.

Thank you very much for your work and patience. I know as a OpenSource contributor sometimes we want to have the code in production, but we care a lot do not to introduce breaking changes to our customers, and this PR of yours is a bit more complex than the others.

**Comment by leandrodamascena** [comment]
Hi @oyiz-michael! After merging the PRs I mentioned in the previous comment, I see we have some merge conflicts here. Can you fix them before we move forward?

Thanks

File: aws_lambda_powertools/event_handler/middlewares/openapi_validation.py [hunk 2]
```diff
@@ -414,6 +597,21 @@ def _handle_missing_field_value(
 
 def _normalize_field_value(value: Any, field_info: FieldInfo) -> Any:
     """Normalize field value, converting lists to single values for non-sequence fields."""
+    # Convert UploadFile to bytes if the expected type is bytes
+    if isinstance(value, UploadFile):
```
Comment (tonnico) [inline aws_lambda_powertools/event_handler/middlewares/openapi_validation.py]: This function was originally meant to normalize values, not convert them. Maybe we should keep conversions separate to avoid changing its intent.

What do you think?

File: aws_lambda_powertools/event_handler/openapi/__init__.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from __future__ import annotations
+
+# OpenAPI module for event handler
+# The fix_upload_file_schema_references function is available in params module
+
+__all__: list[str] = []
```
Comment (tonnico) [inline aws_lambda_powertools/event_handler/openapi/__init__.py]: Looks like we don't need this file.

File: aws_lambda_powertools/event_handler/openapi/params.py [hunk 4]
```diff
@@ -30,6 +30,143 @@
 This turns the low-level function signature into typed, validated Pydantic models for consumption.
 """
 
+__all__ = ["Path", "Query", "Header", "Body", "Form", "File", "UploadFile"]
```
Comment (tonnico) [inline aws_lambda_powertools/event_handler/openapi/params.py:35]: Adding `__all__` here alters what the module exposes. Was the goal to define a stable public surface?

File: aws_lambda_powertools/event_handler/openapi/params.py [hunk 5]
```diff
@@ -1229,3 +1365,247 @@ def _create_model_field(
         required=field_info.default in (Required, Undefined),
         field_info=field_info,
     )
+
+
+def fix_upload_file_schema_references(schema_dict: dict[str, Any]) -> dict[str, Any]:
+    """
+    Fix missing component references for UploadFile in OpenAPI schemas.
```
Comment (tonnico) [inline aws_lambda_powertools/event_handler/openapi/params.py]: “Fix" might not be the right term here. This adds new behavior rather than correcting an existing bug.

File: examples/event_handler/openapi_schema_fix_example.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
+"""
+OpenAPI Schema Fix Example
```
Comment (tonnico) [inline examples/event_handler/openapi_schema_fix_example.py]: This file seems unnecessary and might just be a leftover from development.

File: examples/event_handler/schema_validation_test.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+"""
+OpenAPI Schema Validation Test
```
Comment (tonnico) [inline examples/event_handler/schema_validation_test.py]: This file seems unnecessary and might just be a leftover from development.

File: aws_lambda_powertools/event_handler/middlewares/openapi_validation.py (hunk 2 above)
Comment (oyiz-michael) [reply aws_lambda_powertools/event_handler/middlewares/openapi_validation.py]: I agree. Normalization should stay free of IO/side-effects. I'll move the UploadFile→bytes logic out of the normalizer into a dedicated conversion helper that we call only when the field's expected type requires it (and add tests). If you'd prefer immediate compatibility, I can first add a flag to the normalizer to preserve current behavior and deprecate it in a follow-up. Which approach do you prefer?

File: aws_lambda_powertools/event_handler/openapi/params.py (hunk 4 above)
Comment (oyiz-michael) [reply aws_lambda_powertools/event_handler/openapi/params.py:35]: yes, that was intentional: __all__ is used here to define the stable public surface for openapi params so downstream imports and docs are explicit. I'll add a short comment explaining the intent and keep the list maintained. If you'd rather we leave it implicit, I can remove it instead.

File: aws_lambda_powertools/event_handler/openapi/params.py (hunk 5 above)
Comment (oyiz-michael) [reply aws_lambda_powertools/event_handler/openapi/params.py]: I agree "fix" was misleading; I renamed it to ensure_upload_file_schema_references to reflect that it ensures/augments schema refs rather than correcting a bug.

File: aws_lambda_powertools/event_handler/middlewares/openapi_validation.py (hunk 2 above)
Comment (tonnico) [reply aws_lambda_powertools/event_handler/middlewares/openapi_validation.py]: I wouldn't deprecate that part. It isn't public.

File: aws_lambda_powertools/event_handler/openapi/params.py (hunk 4 above)
Comment (oyiz-michael) [reply aws_lambda_powertools/event_handler/openapi/params.py:35]: @tonnico @leandrodamascena  I've added a comment explaining the intent and will keep it maintained when adding new parameter types. Let me know if you'd prefer a different approach!

File: aws_lambda_powertools/event_handler/middlewares/openapi_validation.py (hunk 2 above)
Comment (oyiz-michael) [reply aws_lambda_powertools/event_handler/middlewares/openapi_validation.py]: Done! I've refactored this in commit 70e8d3aa:

Created a dedicated _convert_uploadfile_to_bytes() function that handles the UploadFile→bytes conversion
Removed the conversion logic from _normalize_field_value() so it only does structural normalization
Both call sites now chain: normalize → convert → validate

File: aws_lambda_powertools/event_handler/middlewares/openapi_validation.py [hunk 8]
```diff
@@ -171,6 +177,142 @@ def _parse_form_data(self, app: EventHandlerInstance) -> dict[str, Any]:
...
+        import base64
+
+        body = app.current_event.body or ""
+
+        if app.current_event.is_base64_encoded:
+            try:
+                return base64.b64decode(body)
+            except Exception:
+                # If decoding fails, use body as-is
+                return body.encode("utf-8") if isinstance(body, str) else body
```
Comment (tonnico) [inline aws_lambda_powertools/event_handler/middlewares/openapi_validation.py]: Is there a case where `is_base64_encoded` is true but it's not base64-encoded?

**Comment by tonnico** [comment]
Would it make sense to rely on [`python-multipart`](https://github.com/Kludex/python-multipart) here instead of implementing custom parsing logic?
That library is already used by Starlette and FastAPI, handles boundary quirks and large uploads reliably, and would likely reduce maintenance and edge-case handling on our side. Using it as an optional dependency might be a cleaner long-term approach.

**Comment by leandrodamascena** [comment]
I'm reviewing this PR this week.

**Comment by oyiz-michael** [comment]
> I'm reviewing this PR this week.

Hi @leandrodamascena  let me know if you require any further changes for this Pr and let me know what the plan is for this Pr. thanks

**Review by dreamorosi — COMMENTED** [review]
Hi @oyiz-michael, apologies for the long delay in reviewing and getting this merged.

The team is quite busy with http://reinvent.awsevents.com prep and because of this we're slightly reprioritizing non-critical PRs.

We'll do our best to review this as soon as possible, but most likely we won't be able to provide a full review before the second week of December, which is after the event.

Thank you for your patience and understanding.

**Comment by oyiz-michael** [comment]
Hi @dreamorosi, @leandrodamascena - hope re:Invent went well!

Following up on this PR now 

Thanks for your patience, and let me know what you'd like me to do next!

**Comment by oyiz-michael** [comment]
Hi @leandrodamascena any feedback on this PR?

**Comment by leandrodamascena** [comment]
Hey @oyiz-michael! First of all, thank you so much for taking the time to work on this - you're an incredible contributor and the effort you put into this PR really shows.

File parameter support is something I really needed, especially because we can now say that OpenAPI Spec is fully implemented here. Your PR helped me understand the full scope and requirements much better. It was honestly the base for my work.

After reviewing the full PR a bunch of time and also trying to fix this with some GenAI help, I realized the scope grew quite a bit over time and it became really hard to review and merge. That's partly on us for taking so long to review and I'm really sorry about the delay.

So I decided to start fresh with a clean implementation that builds directly on this PR. 

I'll give you full credit in the new PR as co-author. Your work made this possible.

Thanks again for the contribution, I really appreciate it!
